src/tools/remote/src/main/java/com/google/devtools/build/remote/worker/CasServer.java

```diff
@@ -71,4 +86,62 @@ public void batchUpdateBlobs(
     responseObserver.onNext(batchResponse.build());
     responseObserver.onCompleted();
   }
+
+  @Override
+  public void getTree(GetTreeRequest request, StreamObserver<GetTreeResponse> responseObserver) {
+    // Directories are returned in depth-first order.  We store all previously-traversed digests so
+    // identical subtrees having the same digest will only be traversed and returned once.
+    Set<Digest> seenDigests = new HashSet<>();
+    Deque<Digest> pendingDigests = new ArrayDeque<>();
+    seenDigests.add(request.getRootDigest());
+    pendingDigests.push(request.getRootDigest());
+    // The page token is implemented as the previous directory's digest.  If the client passes a
+    // page token, we still do the complete traversal internally, but we skip sending the results to
+    // the client until we see the page token's digest.
+    boolean skipping = !request.getPageToken().isEmpty();
+    while (!pendingDigests.isEmpty() && !Context.current().isCancelled()) {
+      Digest digest = pendingDigests.pop();
+      byte[] directoryBytes;
+      try {
+        directoryBytes = getFromFuture(cache.downloadBlob(digest));
+      } catch (CacheNotFoundException e) {
+        responseObserver.onError(StatusUtils.notFoundError(digest));
+        return;
+      } catch (Exception e) {
+        logger.log(WARNING, "Read request failed.", e);
+        responseObserver.onError(StatusUtils.internalError(e));
+        return;
+      }
+      Directory directory;
+      try {
+        directory = Directory.parseFrom(directoryBytes);
+      } catch (InvalidProtocolBufferException e) {
+        logger.log(WARNING, "Failed to parse directory in tree.", e);
+        responseObserver.onError(StatusUtils.internalError(e));
+        return;
+      }
+      for (DirectoryNode directoryNode : directory.getDirectoriesList()) {
+        if (!seenDigests.contains(directoryNode.getDigest())) {
+          seenDigests.add(directoryNode.getDigest());
+          pendingDigests.push(directoryNode.getDigest());
+        }
+      }
+      GetTreeResponse response =
+          GetTreeResponse.newBuilder()
+              .addDirectories(directory)
+              .setNextPageToken(pendingDigests.isEmpty() ? "" : digest.getHash())
```

Comment:
Sure you don't want to just return all of them without page tokens? This will only fail on directories so large that they exceed the gRPC frame size. Probably still fine for testing purposes, but a lot more efficient and easier to write, too.

Reply:
I was concerned that typical trees for actions might go over the limit if e.g. the toolchain is checked in, but it looks like this might not be too large a problem in practice, so I've simplified it in the way you suggest.